cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -180,58 +182,94 @@ public void run() {
+        _stopRequested.set(false);
         _executionTaskManager.clear();
         KafkaCruiseControlUtils.closeZkUtilsWithTimeout(_zkUtils, 10000);
-        _state.set(ExecutorState.State.NO_TASK_IN_PROGRESS);
-        _zkUtils.close();
+        _state = ExecutorState.State.NO_TASK_IN_PROGRESS;
+        // It is possible that the _executorState might be outdated if the user checks it between the two assignments.
+        _executorState = ExecutorState.noTaskInProgress();
       }
     }
 
+    /**
+     * Get the in execution data to move in MB.
+     * @param inExecutionTasks Tasks that are either in progress or aborting.
+     * @return the in execution data to move in MB.
+     */
+    private long inExecutionDataToMoveInMB(Set<ExecutionTask> inExecutionTasks) {
+      return inExecutionTasks.stream().mapToLong(task -> task.proposal().dataToMoveInMB()).sum();
+    }
+
     private void moveReplicas() {
       int numTotalPartitionMovements = _executionTaskManager.remainingPartitionMovements().size();
       long totalDataToMoveInMB = _executionTaskManager.remainingDataToMoveInMB();
-      int partitionsToMove = numTotalPartitionMovements;
       LOG.info("Starting {} partition movements.", numTotalPartitionMovements);
+
+      int partitionsToMove = numTotalPartitionMovements;
       // Exhaust all the pending partition movements.
-      while ((partitionsToMove > 0 || !_executionTaskManager.inExecutionTasks().isEmpty()) && !_stopRequested) {
+      while ((partitionsToMove > 0 || !_executionTaskManager.inExecutionTasks().isEmpty()) && !_stopRequested.get()) {
         // Get tasks to execute.
         List<ExecutionTask> tasksToExecute = _executionTaskManager.getReplicaMovementTasks();
-        LOG.info("Executor will execute " + tasksToExecute.size() + " task(s)");
+        LOG.info("Executor will execute {} task(s)", tasksToExecute.size());
 
         if (!tasksToExecute.isEmpty()) {
           // Execute the tasks.
           _executionTaskManager.markTasksInProgress(tasksToExecute);
           ExecutorUtils.executeReplicaReassignmentTasks(_zkUtils, tasksToExecute);
         }
-        // Wait for some partition movements to finish
+        // Wait indefinitely for partition movements to finish.
         waitForExecutionTaskToFinish();
         partitionsToMove = _executionTaskManager.remainingPartitionMovements().size();
         long dataToMove = _executionTaskManager.remainingDataToMoveInMB();
-        _numFinishedPartitionMovements =
-            numTotalPartitionMovements - partitionsToMove - _executionTaskManager.inExecutionTasks().size();
-        _finishedDataMovementInMB = totalDataToMoveInMB - dataToMove;
+        Set<ExecutionTask> inExecutionTasks = _executionTaskManager.inExecutionTasks();
+        _numFinishedPartitionMovements = numTotalPartitionMovements - partitionsToMove - inExecutionTasks.size();
+        _finishedDataMovementInMB = totalDataToMoveInMB - dataToMove - inExecutionDataToMoveInMB(inExecutionTasks);
```

Comment:
`inExecutionDataToMoveInMB()` may be invoked for N^2 times which could be a waste. Can we just maintain a variable in `ExecutionTaskManager`?